vendor/bam/src/json/request.rs

```diff
@@ -1,44 +1,129 @@
-pub struct Request {
-    #[serde(rename = "type")]
-    _type: String,
-    headers: HashMap<String, JsonValue>,
-    body: JsonValue,
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct UnvalidatedIncomingRequest {
+    #[serde(flatten)]
+    inner: Request,
 }
 
-impl Request {
-    pub fn new(_type: String, headers: HashMap<String, JsonValue>, body: JsonValue) -> Self {
-        Request {
-            _type,
-            headers,
-            body,
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct ValidatedIncomingRequest {
+    #[serde(flatten)]
+    inner: Request,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct OutgoingRequest {
+    #[serde(flatten)]
+    inner: Request,
+}
+
+impl ValidatedIncomingRequest {
+    pub fn header(&self, key: &str) -> Option<&Header> {
+        self.inner.headers.get(key)
+    }
+
+    pub fn take_header(&mut self, key: &str) -> Option<Header> {
+        self.inner.headers.take(key)
+    }
+
+    pub fn take_body_as<B: DeserializeOwned>(self) -> Result<B, serde_json::Error> {
+        self.inner.take_body_as()
+    }
+}
+
+impl OutgoingRequest {
+    pub fn new<T: Into<String>>(request_type: T) -> Self {
+        Self {
+            inner: Request {
+                request_type: request_type.into(),
+                headers: Headers::default(),
+                body: serde_json::Value::Null,
+            },
+        }
+    }
+
+    pub fn with_header(self, key: &str, header: Header) -> Self {
+        let request = self.inner;
+
+        Self {
+            inner: Request {
+                headers: request.headers.with_header(key, header),
+                ..request
+            },
+        }
+    }
+
+    pub fn with_body(self, body: JsonValue) -> Self {
+        Self {
+            inner: Request { body, ..self.inner },
         }
     }
+}
 
-    pub fn get_header<H: DeserializeOwned>(
-        &self,
-        key: &str,
-    ) -> Option<Result<H, serde_json::Error>> {
-        self.headers
-            .get(key)
-            .map(|header| H::deserialize(header.clone()))
+impl UnvalidatedIncomingRequest {
+    pub fn request_type(&self) -> &str {
+        self.inner.request_type.as_str()
     }
 
-    pub fn get_body<B: DeserializeOwned>(&self) -> Option<Result<B, serde_json::Error>> {
-        if self.body.is_null() {
-            return None;
+    pub(crate) fn validate(
+        self,
+        known_headers: &HashSet<String>,
+    ) -> Result<ValidatedIncomingRequest, HeaderErrors> {
+        let (parsed_headers, errors) = self.inner.headers.into_iter().fold(
+            (Headers::default(), HeaderErrors::new()),
+            |(parsed_headers, mut errors), (key, header)| {
+                if !known_headers.contains(&key.value) && key.must_understand {
+                    errors.add_error(key.value.clone(), HeaderErrorKind::UnknownMandatoryHeader)
+                }
+
+                (parsed_headers.with_header(&key.value, header), errors)
```

Comment:
🔧 Maybe don't try and and add the header to parsed_headers if it's errored.

Not that it matters that much but it might make things clearer.